src/python/pants/option/option_types.py

```diff
@@ -0,0 +1,481 @@
+
+
+class OptionBase(Generic[PropType]):
+    """Descriptor for subsystem options.
+
+    Clients shouldn't use this class directly, instead use one of the concrete classes below.
+
+    This class serves two purposes:
+        - Collect registration values for your option.
+        - Provide a typed property for Python usage
+
+    In order to define the `type` registration option, `self.option_type` must return a valid
+    registration type. This can either be set using a class variable or by passing it into
+    `__new__`.
+
+    NOTE: Due to https://github.com/python/mypy/issues/5146 subclasses unfortunately need to provide
+    overloaded `__new__` methods, as subclasses do not inherit overloaded function's annotations.
+    Use one of the existing subclasses as a guide on how to do this.
+    """
+
+    flag_names: tuple[str, ...]
+    kwargs: Dict
+    option_type: Any  # NB: This should be some kind of callable that returns PropType
+
+    # NB: Due to https://github.com/python/mypy/issues/5146, we try to keep the parameter list as
+    # short as possible to avoid having to repeat the param in each of the 3 `__new__` specs that
+    # each subclass must provide.
+    # If you need additional information, this class follows the "Builder" pattern. Use the
+    # "builder" methods to attach additional information. (E.g. see `advanced` method)
+    # NB: We define `__new__` rather than `__init__` because otherwise subclasses would have to
+    # define 3 `__new__`s AND `__init__`.
+    def __new__(
+        cls,
+        *flag_names: str,
+        default: Any,
+        option_type: Type | None = None,
+        help: str,
+    ):
+        self = super().__new__(cls)
+        self.flag_names = flag_names
+        if option_type is None:
+            option_type = cls.option_type
+
+        self.kwargs = dict(
+            type=option_type,
+            default=default,
+            help=help,
+        )
+        return self
+
+    @overload
+    def __get__(self, obj: None, objtype: Any) -> OptionBase[PropType]:
+        ...
+
+    @overload
+    def __get__(self, obj: object, objtype: Any) -> PropType:
+        ...
+
+    def __get__(self, obj, objtype):
+        if obj is None:
+            return self
+        long_name = self.flag_names[-1]
+        option_value = getattr(obj.options, long_name[2:].replace("-", "_"))
+        if option_value is None:
+            return None
+        return self._convert_(option_value)
+
+    # Subclasses can override if necessary
+    def _convert_(self, val: Any) -> PropType:
+        return cast("PropType", self.kwargs["type"](val))
+
+    # ===== "Builder" Methods =====
+    def advanced(self) -> OptionBase[PropType]:
+        self.kwargs["advanced"] = True
+        return self
+
+    def from_file(self) -> OptionBase[PropType]:
+        self.kwargs["fromfile"] = True
+        return self
+
+    def metavar(self, metavar: str, /) -> OptionBase[PropType]:
```

Comment:
This was added in 3.8. Bummer! https://www.python.org/dev/peps/pep-0570/

Reply:
Yep! I really want to be able to upgrade pants.

Reply:
When Pants-embeds-Python. You'll be free to use whatever version :tada: